File: combinatorics/resource/ST/project_5/bai_toan_4/simple_graph/am_to_eal.cpp
#include <iostream>
#include <vector>
#include <utility> // pair

using namespace std;

// Ham chuyen tu ma tran ke co trong so sang danh sach ke mo rong (EAL)
vector<vector<pair<int, int>>> adjacencyMatrixToExtendedList(const vector<vector<int>>& matrix) {
    int n = matrix.size(); // So dinh cua do thi
    vector<vector<pair<int, int>>> adjList(n); // Khoi tao danh sach ke mo rong

    for (int i = 0; i < n; ++i) { // Duyet tung dinh i
        for (int j = 0; j < n; ++j) { // Duyet tung dinh j
            int weight = matrix[i][j]; // Lay trong so canh tu i den j
            if (weight != 0) { // Neu co canh (trong so khac 0)
                adjList[i].emplace_back(j, weight); // Them (j, trong so) vao danh sach ke cua i
            }
        }
    }

    return adjList; // Tra ve danh sach ke mo rong
}

// Ham in danh sach ke mo rong
void printExtendedList(const vector<vector<pair<int, int>>>& adjList) {
    for (int i = 0; i < adjList.size(); ++i) {
        cout << i << ": ";
        for (const auto& [v, w] : adjList[i]) {
            cout << "(" << v << ", " << w << ") ";
        }
        cout << endl;
    }
}

int main() {
    // Ma tran ke co trong so
    vector<vector<int>> adjMatrix = {
        {0, 2, 0, 0},
        {2, 0, 3, 1},
        {0, 3, 0, 4},
        {0, 1, 4, 0}
    };

    // Chuyen doi va in ket qua
    auto adjList = adjacencyMatrixToExtendedList(adjMatrix);
    printExtendedList(adjList);
    return 0;
}

File: combinatorics/resource/ST/project_5/bai_toan_4/simple_graph/am_to_eal.py
# Ham chuyen tu ma tran ke co trong so sang danh sach ke mo rong (EAL)
def adjacency_matrix_to_extended_list(matrix):
    n = len(matrix)  # So dinh cua do thi = so dong cua ma tran
    adj_list = [[] for _ in range(n)]  # Khoi tao danh sach ke rong cho tung dinh

    for i in range(n):  # Duyet tung dinh i
        for j in range(n):  # Duyet tung dinh j
            weight = matrix[i][j]  # Lay trong so canh tu i den j
            if weight != 0:  # Neu co canh (tuc la trong so khac 0)
                adj_list[i].append((j, weight))  # Them (j, trong so) vao danh sach ke cua i

    return adj_list  # Tra ve danh sach ke mo rong

# Ham in danh sach ke mo rong ra man hinh
def print_extended_adj_list(adj_list):
    for i, neighbors in enumerate(adj_list):
        line = ", ".join([f"({v}, {w})" for v, w in neighbors])  # Chuyen danh sach thanh chuoi
        print(f"{i}: {line}")  # In danh sach cua tung dinh

# Vi du: ma tran ke co trong so
adj_matrix = [
    [0, 2, 0, 0],  # Dinh 0 ke voi 1, trong so 2
    [2, 0, 3, 1],  # Dinh 1 ke voi 0, 2, 3
    [0, 3, 0, 4],  # Dinh 2 ke voi 1, 3
    [0, 1, 4, 0]   # Dinh 3 ke voi 1, 2
]

# Chuyen va in danh sach ke mo rong
adj_list = adjacency_matrix_to_extended_list(adj_matrix)
print_extended_adj_list(adj_list)